src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java

```diff
@@ -222,11 +266,18 @@ public MemoryLayout layout() {
     // Layout path construction
 
     public static LayoutPath rootPath(MemoryLayout layout) {
-        return new LayoutPath(layout, 0L, EMPTY_STRIDES, EMPTY_BOUNDS, null);
+        return new LayoutPath(layout, 0L, EMPTY_STRIDES, EMPTY_BOUNDS, EMPTY_DEREF_HANDLES, null);
+    }
+
+    private static LayoutPath nestedPath(MemoryLayout layout, long offset, long[] strides, long[] bounds, MethodHandle[] derefAdapters, LayoutPath encl) {
+        return new LayoutPath(layout, offset, strides, bounds, derefAdapters, encl);
     }
 
-    private static LayoutPath nestedPath(MemoryLayout layout, long offset, long[] strides, long[] bounds, LayoutPath encl) {
-        return new LayoutPath(layout, offset, strides, bounds, encl);
+    private static LayoutPath derefPath(MemoryLayout layout, MethodHandle handle, LayoutPath encl) {
+        MethodHandle[] handles = new MethodHandle[encl.derefAdapters.length + 1];
+        System.arraycopy(encl.derefAdapters, 0, handles, 0, encl.derefAdapters.length);
```

Comment:
This could be simplified using Arrays.copyOf:

```suggestion
        MethodHandle[] handles = Arrays.copyOf(encl.derefAdapters, encl.derefAdapter.length + 1);
```

(no need for separate call to System.arraycopy)